omnipaxos_core/src/sequence_paxos.rs

```diff
@@ -612,16 +624,17 @@ where
                 self.leader_state.majority,
             );
             self.leader = leader_pid;
-            self.storage.set_promise(n);
+            self.storage.set_promise(n).expect("Failed to set promied round");
+            self.cached_state.set_promise(n);
             /* insert my promise */
-            let na = self.storage.get_accepted_round();
-            let ld = self.storage.get_decided_idx();
-            let la = self.storage.get_log_len();
+            let na = self.cached_state.get_accepted_round();
+            let ld = self.cached_state.get_decided_idx();
+            let la = self.storage.get_log_len().expect("Failed to get log length");
             let my_promise = Promise::with(n, na, None, ld, la, self.get_stopsign());
             self.leader_state.set_promise(my_promise, self.pid);
             /* initialise longest chosen sequence and update state */
             self.state = (Role::Leader, Phase::Prepare);
-            let prep = Prepare::with(n, ld, self.storage.get_accepted_round(), la);
+            let prep = Prepare::with(n, ld, self.storage.get_accepted_round().expect("Failed to get accepted round"), la);
```

Comment:
Can't you just read from the cached state? Double check this for all gets.

```suggestion
            let prep = Prepare::with(n, ld, self.cached_state.get_accepted_round().expect("Failed to get accepted round"), la);
```